javalanglib/src/main/scala/java/lang/Double.scala

```diff
@@ -63,22 +63,147 @@ object Double {
-       "([eE][+-]?\\d+)?"+ // optional exponent
-      ")[fFdD]?"         + // optional float / double specifier (ignored)
-      "[\\x00-\\x20]*"   + // optional whitespace
+  private[this] lazy val doubleStrPat = new js.RegExp(
+      "^"                   +
+      "[\\x00-\\x20]*("     + // optional whitespace
+      "[+-]?"               + // optional sign
+      "(?:NaN|Infinity|"    + // special cases
+       "(?:\\d+\\.?\\d*|"   + // literal w/  leading digit
+        "\\.\\d+)"          + // literal w/o leading digit
+       "(?:[eE][+-]?\\d+)?" + // optional exponent
+      ")[fFdD]?"            + // optional float / double specifier (ignored)
+      ")[\\x00-\\x20]*"     + // optional whitespace
+      "$")
+
+  private[this] lazy val doubleStrHexPat = new js.RegExp(
+      "^"                   +
+      "[\\x00-\\x20]*"      + // optional whitespace
+      "([+-]?)"             + // optional sign
+      "0[xX]"               + // hex marker
+      "([0-9A-Fa-f]*)"      + // integral part
+      "\\.?([0-9A-Fa-f]*)"  + // fractional part
+      "[pP]([+-]?\\d+)"     + // binary exponent
+      "[fFdD]?"             + // optional float / double specifier (ignored)
+      "[\\x00-\\x20]*"      + // optional whitespace
       "$")
 
   def parseDouble(s: String): scala.Double = {
-    if (doubleStrPat.test(s))
-      js.Dynamic.global.parseFloat(s).asInstanceOf[scala.Double]
-    else
+    def fail(): Nothing =
       throw new NumberFormatException(s"""For input string: "$s"""")
+
+    // (Very) slow path for hexadecimal notation
+    def parseHexDoubleImpl(match2: js.RegExp.ExecResult): scala.Double = {
+      // scalastyle:off return
+
+      val signStr = match2(1).asInstanceOf[String]
+      val integralPartStr = match2(2).asInstanceOf[String]
+      val fractionalPartStr = match2(3).asInstanceOf[String]
+      val binaryExpStr = match2(4).asInstanceOf[String]
+
+      if (integralPartStr == "" && fractionalPartStr == "")
+        fail()
+
+      def overflow(): scala.Double =
+        if (signStr == "-") scala.Double.NegativeInfinity
+        else scala.Double.PositiveInfinity
+
+      def underflow(): scala.Double =
+        if (signStr == "-") -0.0
+        else 0.0
+
+      /* We concatenate the integral part and fractional part together, then
+       * we parse the result as an integer. This means that we need to remember
+       * a correction to be applied to the final result, as a diff to the
+       * binary exponent
+       */
+      val mantissaStr0 = integralPartStr + fractionalPartStr
+      val correction1 = -(fractionalPartStr.length * 4) // 1 hex == 4 bits
+
+      /* Remove leading 0's in `mantissaStr`, because our algorithm assumes
+       * that there is none.
+       */
+      var i = 0
+      while (i != mantissaStr0.length && mantissaStr0.charAt(i) == '0')
+        i += 1
+      val mantissaStr = mantissaStr0.substring(i)
+      if (mantissaStr == "")
+        return underflow()
```

Comment:
Why is this short exit necessary? Wouldn't it be clearer to truncate `mantissaStr` to at least 1 digit and then follow normal execution flow?

Reply:
Ah, this one is fun. The thing is that if `mantissa == 0.0` makes it to the end, where we multiply it with `correctingPow`, but `correctingPow` is `Infinity` because the binary exponent was too large, then we get `0.0 * Infinity == NaN`. So we need to either explicitly test `mantissa == 0.0` or `correctingPow == Infinty`. Might as well test for the one we find out first.

I have added a comment for that, and a test case that fails if we don't apply the early exit.